src/foam/nanos/ruler/RuleEngine.java

```diff
@@ -51,10 +53,76 @@ public void setDelegate(DAO delegate) {
    * @param oldObj - Old FObject supplied to rules for execution
    */
   public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
-    applyRules(rules, obj, oldObj);
+    CompoundContextAgency compoundAgency = new CompoundContextAgency();
+    ContextualizingAgency agency = new ContextualizingAgency(compoundAgency, userX_, getX());
+    for (Rule rule : rules) {
+      if ( stops_.get() ) break;
+      if ( ! isRuleApplicable(rule, obj, oldObj)) continue;
+      PM pm = new PM();
+      pm.setClassType(RulerDAO.getOwnClassInfo());
+      pm.setName(rule.getName());
+      pm.init_();
+      applyRule(rule, obj, oldObj, agency);
+      pm.log(x_);
+      agency.submit(x_, x -> saveHistory(rule, obj));
+    }
+    compoundAgency.execute(x_);
+    try {
+      //compoundAgency.execute(x_);
+    } catch (Exception e) {
+      Logger logger = (Logger) x_.get("logger");
+      logger.error(e.getMessage());
+    }
+
     asyncApplyRules(rules, obj, oldObj);
   }
 
+  /**
+   * Probes rules execution by applying actions and skipping
+   * execution of agents that contain code that effects the system
+   *
+   * @param rules - Rules to be considered applying
+   * @param obj - FObject supplied to rules for execution
+   * @param rulerProbe -
+   * @param oldObj - Old FObject supplied to rules for execution
+   */
+  public void probe(List<Rule> rules, RulerProbe rulerProbe, FObject obj, FObject oldObj) {
+    PM pm = new PM();
+      pm.setClassType(RulerProbe.getOwnClassInfo());
+      pm.setName("Ruler Probe on " + obj.getClassInfo() + ". id: " + obj.getProperty("id"));
```

Comment:
The name is meant to be brief so it can appear in the the Performance tableView.
Just:
pm.setName("Probe:" + obj.getClassInfo());
I don't think we need the id, since otherwise we would have too many PM's: one for each object.